admin/aws.py

```diff
@@ -187,18 +189,158 @@ def perform_list_s3_keys(dispatcher, intent):
             for key in bucket.list(intent.prefix)}
 
 
+@attributes([
+    "source_bucket",
+    "source_prefix",
+    "target_path",
+    "filter_extensions",
+])
+class DownloadS3KeyRecursively(object):
+    """
+    Download the S3 files from a key in a bucket to a given directory.
+
+    :ivar bytes source_bucket: Name of bucket to download keys from.
+    :ivar bytes source_prefix: Download only files with this prefix.
+    :ivar FilePath target_path: Directory to download files to.
+    :ivar tuple filter_extensions: Download only files with extensions in this
+        tuple.
+    """
+
+
+@sync_performer
+@do
+def perform_download_s3_key_recursively(dispatcher, intent):
+    """
+    See :class:`DownloadS3KeyRecursively`.
+    """
+    keys = yield Effect(
+        ListS3Keys(prefix=intent.source_prefix + '/',
+                   bucket=intent.source_bucket))
+    for key in keys:
+        if not key.endswith(intent.filter_extensions):
+            continue
+        path = intent.target_path.preauthChild(key)
+
+        if not path.parent().exists():
+            path.parent().makedirs()
+        source_key = os.path.join(intent.source_prefix, key)
+        yield Effect(
+            DownloadS3Key(source_bucket=intent.source_bucket,
+                          source_key=source_key,
+                          target_path=path))
+
+
+@attributes([
+    "source_bucket",
+    "source_key",
+    "target_path",
+])
+class DownloadS3Key(object):
+    """
+    Download a file from S3.
+
+    :ivar bytes source_bucket: Name of bucket to download key from.
+    :ivar bytes source_key: Name of key to download.
+    :ivar FilePath target_path: Path to download file to.
+    """
+
+
+@sync_performer
+def perform_download_s3_key(dispatcher, intent):
+    """
+    See :class:`DownloadS3Key`.
+    """
+    s3 = boto.connect_s3()
+
+    bucket = s3.get_bucket(intent.source_bucket)
+    key = bucket.get_key(intent.source_key)
+    with intent.target_path.open('w') as target_file:
+        key.get_contents_to_file(target_file)
+
+
+@attributes([
+    "source_path",
+    "target_bucket",
+    "target_key",
+    "files",
+])
+class UploadToS3Recursively(object):
+    """
+    Upload contents of a directory to S3, for given files.
+
+    Note that this returns a list with the prefixes stripped.
+
+    :ivar FilePath source_path: Prefix of files to be uploaded.
+    :ivar bytes target_bucket: Name of bucket to upload file to.
+    :ivar bytes target_key: Name S3 key to upload file to.
+    :ivar list files: List of bytes, filenames to upload.
+    """
+
+
+@sync_performer
+@do
+def perform_upload_s3_key_recursively(dispatcher, intent):
+    """
+    See :class:`UploadToS3Recursively`.
+    """
+    for path in intent.source_path.walk():
+        if path.basename() in intent.files:
```

Comment:
If you have the list of files, you don't actually need to walk the tree. In any case, `files` should be a list of paths relative to `source_path`, so this won't properly grab files in subdirectories.